src/job.rs

```diff
@@ -13,9 +14,35 @@ impl From<u64> for Id {
 }
 
 impl From<&glue::Job> for Id {
-    fn from(job: &glue::Job) -> Self {
+    /// We don't care about order in some places (e.g. output file) while we do
+    /// in others (e.g. command arguments.) The hash should reflect this!
+    ///
+    /// Note: this data structure is going to grow the ability to refer to other
+    /// jobs as soon as it's feasible. When that happens, a depth-first search
+    /// through the tree rooted at `top_job` will probably suffice.
+    fn from(top_job: &glue::Job) -> Self {
+        // TODO: is this the best hash for this kind of data? Should we find
+        // a faster one?
         let mut hasher = std::collections::hash_map::DefaultHasher::new();
-        job.hash(&mut hasher);
+
+        let job = &top_job.f0;
+
+        // TODO: when we can get commands from other jobs, we need to hash the
+        // other tool and job instead of relying on the derived `Hash` trait
+        // for this.
+        job.command.hash(&mut hasher);
+
+        // TODO: input file hashes need to change this hash. We cannot do that
```

Comment:
we need to at least know file metadata for inputs in advance (so we know when to re-run the build.) We also need to get the hashes (so we know when metadata changes don't indicate actual changes.) That's out of scope here, but happened in #37.

We don't need to do the same things for output files because they cannot be known in advance (but this PR takes care of storing them in a content-addressable way.)

Reply:
Ack.